octodns/provider/azuredns.py

```diff
@@ -805,23 +841,58 @@ def _data_for_dynamic(self, azrecord):
                 'pools': pools,
                 'rules': rules,
             },
-            'value': _check_endswith_dot(default[0]),
         }
 
+        if typ == 'CNAME':
+            data['value'] = _check_endswith_dot(default[0])
+        else:
+            data['values'] = default
+
         return data
 
     def _extra_changes(self, existing, desired, changes):
         changed = set()
 
-        # Abort if there are non-CNAME dynamic records
+        # Abort if there are unsupported dynamic records
         for change in changes:
             record = change.record
             changed.add(record)
             typ = record._type
             dynamic = getattr(record, 'dynamic', False)
-            if dynamic and typ != 'CNAME':
-                msg = '{}: Dynamic records in Azure must be of type CNAME'
-                msg = msg.format(record.fqdn)
+            if not dynamic:
+                continue
+            if typ in ['A', 'AAAA']:
+                data = dynamic._data()
+                values = set(record.values)
+                pools = data['pools'].values()
+                seen_values = set()
+                rr = False
+                fallback = False
+                for pool in pools:
+                    vals = pool['values']
+                    if len(vals) > 1:
+                        rr = True
+                    pool_values = set(val['value'] for val in vals)
+                    if pool.get('fallback'):
+                        fallback = True
+                    seen_values.update(pool_values)
+
+                if values != seen_values:
+                    msg = ('{} {}: All pool values of A/AAAA dynamic records '
+                           'must be included in top-level \'values\'.'
+                           .format(record.fqdn, record._type))
+                    raise AzureException(msg)
+
+                geo = any(r.get('geos') for r in data['rules'])
+
+                if [rr, fallback, geo].count(True) > 1:
+                    msg = ('{} {}: A/AAAA dynamic records must use at most '
+                           'one of round-robin, fallback and geo-fencing')
+                    msg = msg.format(record.fqdn, record._type)
+                    raise AzureException(msg)
+            elif typ != 'CNAME':
+                msg = ('{}: Dynamic records in Azure must be of type '
+                       'A/AAAA/CNAME').format(record.fqdn)
```

Comment:
This method is getting pretty long. May be wroth breaking up some of these branches/bits into helper functions to aid readability. Doesn't have to happen right now if you don't want. It can be in a dedicated PR or something if you'd prefer (if done at all.)

Reply:
👍 good call. updated the PR.